verification/certifyKey.go

```diff
@@ -483,114 +83,259 @@ func testCertifyKey(d TestDPEInstance, c DPEClient, t *testing.T, simulation boo
 		// Check key returned in command response against certificate
 		checkCertifyKeyResponse(t, leafCert, *certifyKeyResp, hashAlg)
 
-		// Validate that all X.509 fields conform with the format defined in the DPE iRoT profile
-		validateCertifyKeyCert(t, leafCert, params.Flags, params.Label)
-
 		// Ensure full certificate chain has valid signatures
 		// This also checks certificate lifetime, signatures as part of cert chain validation
-		validateLeafCertChain(t, certChain, leafCert)
+		if err = validateLeafCertChain(certChain, leafCert); err != nil {
+			t.Errorf("[ERROR]: %v", err)
+		}
+
+		// Check for basic constraints extension
+		if err = checkBasicConstraints(leafCert, params.Flags); err != nil {
+			t.Errorf("[ERROR]: %v", err)
+		}
+
+		// Check key usage extensions - DigitalSignature, CertSign
+		if err = checkKeyExtensions(leafCert); err != nil {
+			t.Errorf("[ERROR]: %v", err)
+		}
+
+		// Check extended key usage extensions
+		if err = checkExtendedKeyUsages(leafCert); err != nil {
+			t.Errorf("[ERROR]: %v", err)
+		}
+
+		// Check critical UEID and Multi Tcb Info TCG extensions
+		// Check UEID extension
+		if err = checkTcgUeidExtension(leafCert, params.Label); err != nil {
+			t.Errorf("[ERROR]: %v", err)
+		}
+
+		// Check MultiTcbInfo Extension structure
+		if _, err = getMultiTcbInfo(leafCert); err != nil {
+			t.Errorf("[ERROR]: %v", err)
+		}
 
 		// Reassign handle for simulation mode.
 		// However, this does not impact in default mode because
 		// same default context handle is returned in default mode.
 		handle = &certifyKeyResp.Handle
 	}
-	// TODO: When DeriveChild is implemented, call it here to add more TCIs and call CertifyKey again.
+	// DeriveChild to add more TCIs and call CertifyKey again.
+	if simulation {
+		checkWithDerivedChildContextSimulation(d, c, t, handle)
+	} else {
+		checkWithDerivedChildContext(d, c, t, handle)
+	}
```

Comment:
Must have two separate methods to test derived child contexts as clubbing together lead to confusing if-elses.
The simulation mode varies in locality, Derivechild flags, not needing restoration of parent handle etc...